core/src/main/java/com/linecorp/armeria/internal/common/util/KeyStoreUtil.java

```diff
@@ -0,0 +1,186 @@
+                               @Nullable String keyStorePassword,
+                               @Nullable String keyPassword,
+                               @Nullable String alias) throws IOException, GeneralSecurityException {
+        try (InputStream in = new FileInputStream(keyStoreFile)) {
+            return load(in, keyStorePassword, keyPassword, alias, keyStoreFile);
+        }
+    }
+
+    public static KeyPair load(InputStream keyStoreStream,
+                               @Nullable String keyStorePassword,
+                               @Nullable String keyPassword,
+                               @Nullable String alias) throws IOException, GeneralSecurityException {
+        return load(keyStoreStream, keyStorePassword, keyPassword, alias, null);
+    }
+
+    private static KeyPair load(InputStream keyStoreStream,
+                                @Nullable String keyStorePassword,
+                                @Nullable String keyPassword,
+                                @Nullable String alias,
+                                @Nullable File keyStoreFile)
+            throws IOException, GeneralSecurityException {
+
+        try (InputStream in = new BufferedInputStream(keyStoreStream, 8192)) {
+            in.mark(4);
+            final String format = detectFormat(in);
+            in.reset();
+
+            if (format == null) {
+                throw newException("unknown key store format", keyStoreFile,
+                                   "(expected: PKCS#12 or JKS format)");
+            }
+
+            final KeyStore ks = KeyStore.getInstance(format);
+            ks.load(in, keyStorePassword != null ? keyStorePassword.toCharArray() : null);
+
+            // Referred to Vert.x `KeyStoreHelper` for the logic for discovering a key pair:
+            // https://github.com/eclipse-vertx/vert.x/blob/52f95ab88f0165c7c0db2ba48e4c06180a8d6655/src/main/java/io/vertx/core/net/impl/KeyStoreHelper.java
+            PrivateKey privateKey = null;
+            List<X509Certificate> certificateChain = null;
+            for (final Enumeration<String> e = ks.aliases(); e.hasMoreElements();) {
+                final String candidateAlias = e.nextElement();
+                if (alias != null && !alias.equals(candidateAlias)) {
+                    continue;
+                }
+
+                if (ks.isKeyEntry(candidateAlias)) {
+                    // Extract the private key.
+                    final PrivateKey candidateKey = (PrivateKey) ks.getKey(
+                            candidateAlias, keyPassword(keyStorePassword, keyPassword));
+
+                    // Extract the certificate chain.
+                    final Certificate[] candidateCertificateChain = ks.getCertificateChain(candidateAlias);
+                    if (candidateCertificateChain == null || candidateCertificateChain.length == 0) {
+                        throw newException("the key pair contains no certificate chain", keyStoreFile,
```

Comment:
How about also adding the `alias` (if it's not null) or `candidateAlias` to the message?
`"the key store contains no certificate chain for " + (alias != null ? alias : candidateAlias)`